[astro] Fix offsets if "earliest" and "latest" is null or default "00:00"

pacive opened the pull request:
Fixes #19694

<!--
Thanks for contributing to the openHAB project!
Please describe the goal and effect of your PR here.
Pay attention to the below notes and to *the guidelines* for this repository.
Feel free to delete any comment sections in the template (starting with "<!--").

ATTENTION: Don't use "git merge" when working with your pull request branch!
This can clutter your Git history and make your PR unusable.
Use "git rebase" instead. See this forum post for further details:
https://community.openhab.org/t/rebase-your-code-or-how-to-fix-your-git-history-before-requesting-a-pull/129358

All PRs should be created using the "main" branch as base.
Important bugfixes are cherry-picked by maintainers to the patch release branch after a PR has been merged.

Add one or more appropriate labels to make your PR show up in the release notes.
E.g. enhancement, bug, documentation, new binding
This can only be done by yourself if you already contributed to this repo.

If your PR's code is not backward compatible with previous releases (which
should be avoided), add a message to the release notes by filing another PR:
https://github.com/openhab/openhab-distro/blob/main/distributions/openhab/src/main/resources/bin/update.lst

# Title

Provide a short summary in the *Title* above. It will show up in the release notes.
For example:
- [homematic] Improve communication with weak signal devices
- [timemachine][WIP] Initial contribution

# Description

Please give a few sentences describing the overall goals of the pull request.
Give enough details to make the improvement and changes of the PR understandable
to both developers and tech-savy users.

Please keep the following in mind:
- What is the classification of the PR, e.g. Bugfix, Improvement, Novel Addition, ... ?
- Did you describe the PRs motivation and goal?
- Did you provide a link to any prior discussion, e.g. an issue or community forum thread?
- Did you describe new features for the end user?
- Did you describe any noteworthy changes in usage for the end user?
- Was the documentation updated accordingly, e.g. the add-on README?
- Does your contribution follow the coding guidelines:
  https://www.openhab.org/docs/developer/guidelines.html
- Did you check for any (relevant) issues from the static code analysis:
  https://www.openhab.org/docs/developer/bindings/#include-the-binding-in-the-build
- Did you sign-off your work:
  https://www.openhab.org/docs/developer/contributing.html#sign-your-work

# Testing

Your pull request will automatically be built and available under the following folder:
https://openhab.jfrog.io/ui/native/libs-pullrequest-local/org/openhab/addons/bundles/

It is a good practice to add a URL to your built JAR in this pull request description,
so it is easier for the community to test your Add-on.
If your pull request contains a new binding, it will likely take some time
before it is reviewed and processed by maintainers.
That said, consider submitting your Add-on in the Marketplace:
https://community.openhab.org/c/marketplace/69

Don't forget to submit a thread about your Add-on in the openHAB community:
https://community.openhab.org/c/add-ons 
-->

Nadahar commented:
The changes to these two methods in #19310 seem to have been:
```diff
@@ -205,11 +213,11 @@ public class DateTimeUtils {
     }
 
     public static Calendar getAdjustedEarliest(Calendar cal, AstroChannelConfig config) {
-        return adjustTime(cal, getMinutesFromTime(config.earliest));
+        return config.earliest == null ? cal : adjustTime(cal, getMinutesFromTime(config.earliest));
     }
 
     public static Calendar getAdjustedLatest(Calendar cal, AstroChannelConfig config) {
-        return adjustTime(cal, getMinutesFromTime(config.latest));
+        return config.latest == null ? cal : adjustTime(cal, getMinutesFromTime(config.latest));
     }
```
So, I guess it boils down to what `getMinutesFromTime()` used to do when it was handed `null`?

pacive commented:
The specific changes that altered the behaviour was the changes to `public static Calendar applyConfig(Calendar cal, AstroChannelConfig config)` in `DateTimeUtils.java`. The calls to `getAdjustedEarliest` and `getAdjustedLatest` were changed to use the original `cal` object instead of the `cCal` object that had the offset already applied.

When I just changed those lines back, some tests failed however, so I figured there was some reason for this change, and I changed the implementations of `getAdjustedEarliest` and `getAdjustedLatest` to check the configuration and only adjust the time if not null and not "00:00"

Nadahar commented:
Your implementation applies "earliest" and "latest" even if they aren't set (in setting it to the start or end of the current date). I can't wrap my head around why that is the correct thing to do. If these aren't set, shouldn't `cal` just be returned unaltered?

Nadahar commented:
It also strikes me that this is about how the logic is meant to be. Basically, should earliest/latest or offset "have the last word"? Should an offset be applied which can take the time outside earliest/latest, or should those be absolutes. I guess my thinking at the time was that the limits are absolutes, so if the offset take you outside these, the offset becomes irrelevant.

pacive commented:
After thinking a bit more I agree. The way I see it it should be:
1. Latest/earliest set: offset should be applied, but not in a way that goes outside the bounds.
2. Latest/earliest NOT set: set earliest/latest to start/end of the day; goto 1

Need to look at my fix again to see if that's the behaviour it produces.

pacive commented:
Yes, it currently only works if earliest is set to 00:01. It would be easier if only `null` needed to be handled, but unfortunately, if configuring an offset in MainUI, both earliest and latest will are set to 00:00 by default and that gets saved even if those fields aren't touched, so for ease of use those need to be treated the same. Otherwise the only way would be to go into the code tab and remove the parameters.

Nadahar commented:
> I guess we need some input from @openhab/maintainers on what the expected behaviour should be WRT crossing date boundaries, I can see both ways as equally valid.

The question is if anybody actually has a stance on this. I think historic behavior might be the most relevant guide to use.

pacive commented:
Thanks, will test it out

Nadahar commented:
> `threshold` and `actual` doesn't get initialized if minutes <= 0 (which it it if config.earliest is null.

No, but they aren't used if that's the case either...?

pacive commented:
Sorry, must have been tires and misread the else branch to be for the outer if.

I noticed also that it uses get instance, but what I find strange is that all the other test cases work anyway, and my tests was at least deterministic before I applied your patch.

My guess is that "force" is meant to fire the event at the earliest/latest time configured even if the event would not happen (e.g., sunrise when there's midnight sun). This was one of the reasons I felt setting the bounds to start/end of day was a reasonable approach.

pacive commented:
I guess in practice this would mean that both start and end events would fire the moment the job is scheduled by the binding (+offset), which should be close to midnight. Not sure how negative offsets would work here, but I guess they would be fired immediately.

Nadahar commented:
> I noticed also that it uses get instance, but what I find strange is that all the other test cases work anyway, and my tests was at least deterministic before I applied your patch.

That isn't so strange, because you made the "adjust" methods default to "the extremes" if no limits were defined. But, this wasn't the behavior in the past, and the "adjust" methods are also used by other code (and are public), which is why I think it's a bad idea to do.

That said, we can make it do what we want, my problem is that I don't understand "the intention" - what exactly does it mean to "force" a range.

I never dug into all the details in the logic of the binding, I just tried to apply bug fixes to the logic that was already there. Unfortunately, that's not always possible, because fixing one bug just reveals another. But, it's still not clear to me what the consequences of setting the threshold when "forcing" an open-ended range is. If that means that the range should start at 0:00 and end at 23:59:59.999 - that's easy enough to achieve - I just need to understand properly what the goal is, to be able to tell what "the right behavior" should be.

pacive commented:
I've been thinking about the ranges and "force" setting a bit more, and I'm not sure I understand the logic behind how it's implemented. An example just to demonstrate some weirdness:

Let's say I want to turn on some lights when the sun goes down, so I create a rule that triggers on the set START event. But I would like the lights to turn on earliest at 16:00 and 21:00 at the latest, so I configure those as well. Now it happens that I live in the far north of Sweden, so during summer, the sun is up all day around and there is no sunset. This causes the range [START..END] that defines the triggers to be [null..null]. If I enable the "force" setting, when the `adjustRangeToConfig` function is called it will set the range to [16:00..21:00], which means that the START event will be triggered at 16:00. If I instead choose to trigger on the END event, the opposite will happen during winter, as soon as there no longer is a sunset, the rule will be triggered at 21:00, when the day before triggered at 16:00 (since sunset would be close to noon, so it would be adjusted to "earliest").

If I instead would opt to use the "daylight#event" channel, I would of course only be able to use the END event (the START would be for turning off the lights in the morning). Then I can only set the "latest" parameter to restrict the time of the events. Daylight END would become closer to noon during the fall, until there is no daylight anymore, and at that day the event will suddenly jump to trigger at 21:00.

I honestly can't see a valid use case for forcing the events even when null that wouldn't cause similar issues.

Nadahar commented:
> so the only thing we need to worry about is to ensure the offset is applied, without bounds if not earliest/latest is configured, and restrict it to those bounds if they are.

I believe that is working with the code I have suggested. The real bug here was that the offset was never applied, because either the "before" or the "after" check would always trigger. That isn't the case anymore, and I've written tests that validate this behavior.
